Optional compatibility with other Select2 v3 plugins

brandomeniconi opened the pull request:
As discussed in #683 and #670, actually many popular plugin, like Advanced Custom Fields and WooCommerce, still use the Select2 v3 library and make this plugin throw JS errors due to v3/v4 incompatibility.

With this PR the user can: 

`define('SELECT2_NOCONFLICT', true);` in `wp-config.php` 

to make this plugin load select2 4.x in `$.fn.select2v4` namespace, leaving the `$.fn.select2` on v3 for all other plugins. The plugin's JS than uses the `select2v4 ` namespace if exists, otherwise it uses the standard `select2`.

An optional improvement to this PR could be to check if known incompatible plugins  are active with `is_plugin_active()` and automatically set  `SELECT2_NOCONFLICT` to true.

Commit:
Implements optional Select2 version compatibility

Commit:
fixed linting errors

goldenapples commented on js/build/shortcode-ui.js:
```diff
@@ -1768,7 +1768,8 @@ sui.views.editAttributeSelect2Field = sui.views.editAttributeField.extend( {
 		this.preselect( $field );
 
-		var $fieldSelect2 = $field.select2({
+                var select2 = jQuery().select2v4 ? 'select2v4' : 'select2';
```
We could probably pass this handle in along with the localized script variables, so that we don't need to check which name it exists under here.

goldenapples commented on inc/fields/class-field-post-select.php:
```diff
@@ -38,8 +38,10 @@ public function filter_shortcode_ui_fields( $fields ) {
+		$noconflict = defined( 'SELECT2_NOCONFLICT' ) && SELECT2_NOCONFLICT;
+
+		wp_enqueue_script( $noconflict ? 'select2v4' : 'select2' );
+		wp_enqueue_style( $noconflict ? 'select2v4' : 'select2' );
```
One thing that I think might help clean up the code would be to define the Select2 handle in use as a public static variable on the `Shortcode_UI` class. This would get around the need to repeat this pattern of checking the constant in four different places.

goldenapples reviewed (changes requested):
Nice, and thanks for the PR! I've tested this out with Woocommerce and can confirm that the select2 functionality from both plugins work as expected. I was concerned that there would also be style conflicts, but it looks as though the classes used have changed enough from select2 v3 to v4 that there aren't any major clashes from loading two separate stylesheets.

> An optional improvement to this PR could be to check if known incompatible plugins are active with `is_plugin_active()` and automatically set `SELECT2_NOCONFLICT` to true.

Yeah, that could work, but I'd prefer not adding that logic right now. This PR should be a good enough functional fix for site owners who come across this conflict, and trying to add magic conflict detection would be adding one more thing to maintain to a plugin which is already undermaintained. We should just add some clearer documentation so that people who encounter this problem don't have to search through issues here to find the solution.

The failing lint has to do with [the relatively new WPCS file name rule](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/806), and doesn't need to be fixed here. I'll open a separate PR to address that.

Commit:
Set Select2 handle via Shortcode_UI static property

Commit:
Better Select2 handle management

goldenapples commented on inc/class-shortcode-ui.php:
```diff
@@ -199,12 +207,17 @@ public function action_admin_enqueue_scripts( $editor_supports ) {
-		wp_register_style( 'select2',
+		if( self::$select2_handle !== 'select2' ){
+			 wp_add_inline_script( self::$select2_handle, 'var existingSelect2 = jQuery.fn.select2 || null; if (existingSelect2) { delete jQuery.fn.select2; }', 'before' );
+			 wp_add_inline_script( self::$select2_handle, 'jQuery.fn.'.self::$select2_handle.' = jQuery.fn.select2; if (existingSelect2) { delete jQuery.fn.select2; jQuery.fn.select2 = existingSelect2; }', 'after' );
```
I suppose the inline variable should be escaped, lest another plugin get us to run some arbitrary javascript here.

goldenapples replied:
... or, couldn't we just use the handle we've passed in with the localization data? Would `jQuery.fn[ shortcodeUIData.select2_handle ]` work here?

brandomeniconi replied:
Yes of course, great idea and it works! Wordpress loads the localization variables before all scripts, so even if the `shortcodeUIData` is a "translation var" for the `shortcode-ui` script, should be also available to the Select2 script. With that the escaping should be automatically done, right?

goldenapples replied:
> With that the escaping should be automatically done, right?

Yeah, I think the escaping that's being done to prepare the localization strings for json encoding is safe enough. It may not actually guarantee a valid attribute name, but it will prevent passing something like `);alert('xss')` in and running that arbitrary script, which was my concern.

goldenapples reviewed (commented):
Other than my comment about escaping the variable name we pass into the inline script I like this PR, and it seems very useful for allowing compatability with other plugins. @mattheu @danielbachhuber any objections to merging this so we can cut a version that works with the other plugins that we're conflicting with now? 

I have a separate PR in #703 that fixes our linting issues. I'll have to freshen it up after this is merged, but that should get the build passing again.

Commit:
Use shortcodeUIData to set Select2 library handle

Uses JS shortcodeUIData.select2_handle (localization var), instead of direct PHP output, to
set Select2 library handle.

This also prevents potential JS code injection. Ref #702
